### PR TITLE
feat: add plugin configuration validation to lint

### PR DESCRIPTION
**Pull request by jarvis9443**
## Summary

Add plugin configuration validation to ADC's lint functionality. Previously, lint only validated core resource structure (routes, services, upstreams, etc.) via Zod schemas, but accepted any plugin configuration without validation.

### Two-Phase Validation

- **Phase 1 (local)**: Existing Zod schema validation for core resources (unchanged behavior)
- **Phase 2 (optional)**: Plugin config validation using JSON Schema fetched from APISIX/API7 EE Admin API via `GET /apisix/admin/schema/plugins/{name}`

### New Features

- `adc lint` now accepts optional `--backend`, `--server`, `--token`, `--gateway-group` flags to enable plugin validation
- Without backend flags, lint behaves exactly as before (pure local validation)
- `adc diff` and `adc sync` automatically validate plugin configs using the connected backend

### Plugin Validation Coverage

All plugin locations in the configuration are validated:
- `services[*].plugins`, `routes[*].plugins`, `stream_routes[*].plugins`
- `consumers[*].plugins`, `consumer_groups[*].plugins`  
- `consumers[*].credentials[*].config` (using `consumerSchema`)
- `global_rules`, `plugin_metadata` (using `metadataSchema`)

### Backend Support

| Backend | Plugin Validation |
|---|---|
| APISIX | ✅ Fetches schemas via Admin API |
| API7 EE | ✅ Fetches schemas via Admin API |
| APISIX Standalone | ⏭️ Gracefully skipped (no network) |

### Implementation Details

- SDK: Added `PluginSchemaEntry`, `PluginSchemaMap` types and `fetchPluginSchemas()` to Backend interface
- Backends: Implemented `fetchPluginSchemas()` in both BackendAPISIX and BackendAPI7
- Validator: Uses AJV with draft-04 support (matching APISIX JSON Schema format)
- Error reporting: Unified `LintError`/`LintResult` types for both Zod and AJV errors
- Lint command remains a `BaseCommand` (no default backend connection)
- `.gitignore` fix: Changed `/out-tsc` to `out-tsc` to ignore build artifacts in all subdirectories

### Tests

Added 20 new test cases covering:
- Unknown plugin detection
- Missing required fields, type mismatches, multiple errors
- All plugin location traversals (services, routes, consumers, credentials, global_rules, metadata)
- Edge cases (empty plugins, no schemas, core validation priority)

All 45 tests pass (25 existing + 20 new).

## Summary by CodeRabbit

* **New Features**
  * Plugin schema validation added and enabled as an option for lint/sync, plus new CLI flags to configure backend-backed validation.
  * Backend schema fetching with caching to retrieve plugin schemas at runtime.

* **Improvements**
  * Linter now returns structured validation results (success, errors, data) and reports path/code details.
  * Tests updated/expanded for plugin validation coverage and error-shape changes.

### CODE REVIEW

File: apps/cli/src/linter/index.ts [hunk 1]
```diff
@@ -1,6 +1,51 @@
...
+
+export type { LintError, LintOptions, LintResult } from './types';
+
+const zodIssuesToLintErrors = (issues: z.ZodIssue[]): LintError[] =>
+  issues.map(({ path, message, code, ...rest }) => ({
+    path,
+    message,
+    code,
+    ...rest,
+  }));
```
[the remark this answers is not in the document]
Comment (nic-6443) [reply apps/cli/src/linter/index.ts:16]: Already fixed in commit ec31ac2 by changing `LintError.path` type from `(string | number)[]` to `PropertyKey[]` to match Zod's issue path type. CI typecheck now passes.

File: package.json [hunk 2]
```diff
@@ -37,5 +37,9 @@
       "@parcel/watcher",
       "nx"
     ]
+  },
+  "dependencies": {
+    "ajv": "catalog:",
+    "ajv-draft-04": "catalog:"
```
[the remark this answers is not in the document]
Comment (nic-6443) [reply package.json]: Good catch. The `ajv` and `ajv-draft-04` entries in root `package.json` were mistakenly added - they should only be in the workspace catalog (`pnpm-workspace.yaml`) and `apps/cli/package.json`. The `ajv` direct import was already removed from cli deps (only `ajv-draft-04` is imported). Removing the root package.json entries now.

Note: `ajv-draft-04` IS directly imported in `apps/cli/src/linter/plugin-validator.ts` - the search script may have missed it due to file type filtering.

**Comment by nic-6443** [comment]
Addressed CodeRabbit review findings:

1. **FetchPluginSchemasTask gated by --no-lint** (sync.command.ts, diff.command.ts): Fixed. Now uses `opts.lint ? FetchPluginSchemasTask() : { task: () => undefined }` pattern, consistent with how LintTask is already gated.

2. **diff.command.ts catch(err) swallowing**: This is pre-existing code not introduced by this PR - will not change it here to keep the diff focused.

Nitpick comments noted but not applied to keep the PR scope focused on the core plugin validation feature.

**Comment by bzp2010** [comment]
This does not solve 90% of the problems, because configuration errors outside of any plugins cannot be checked. As a result, even if the advanced mode of the linter passes, the false negatives it reports are meaningless, since the server will reject other configuration errors.

Furthermore, even if we take a step back and assume that plugin-only validation is acceptable, the current approach cannot verify whether the plugin configuration is truly, completely correct. This is because many plugins include not only JSON Schema validation but also code-driven configuration validation, which the current method cannot check. A prime example is the CORS plugin.
Even if the lint passes, the server may still reject the configuration at any time, meaning that the lint cannot guarantee anything.

And without a once-and-for-all solution, I am unwilling to endorse the functionality claimed by the lint tool.

---

Even if this PR doesn’t exist, we can still prevent configuration errors through workflow design.
For example, we can synchronize changes to the test environment when a commit is pushed to a PR, meaning that modifications in the current sync are first applied to the test environment, thereby avoiding any impact on the production environment.

Once the PR is merged or a tag is pushed to the main branch, we can then initiate a sync to the production environment—all this can be achieved simply by loading different environment variables at runtime.

Any configuration errors are detected and fixed during the PR before the PR is merged. As long as all CI checks pass, the configuration can be deployed to the production environment without issues.

---

In fact, the `lint` subcommand has been intentionally designed to operate in pure local mode since its creation.
We have always reserved the `validate` subcommand, which is intended to perform server-side configuration validation. This is what I hope to add for the apisix-standalone backend, as it will facilitate pre-checks during AIC CRD submissions.
I believe this is the only way to effectively address all the configuration validation issues we aim to resolve. By simply running `adc lint` + `adc validate` + `adc sync`, users can ensure their configuration is always can be accepted.

**Comment by bzp2010** [comment]
Overall, I’m not optimistic about the content added in this PR.

**Comment by nic-6443** [comment]
Thanks for the detailed feedback @bzp2010. These are important architectural points worth discussing.

**Regarding the limitations you raised:**

You're right that JSON Schema alone can't catch everything — code-driven validation (like CORS's custom checks) is beyond what schema validation can cover. And non-plugin config errors are already handled by the existing Zod validation in Phase 1, but the server-side checks go deeper.

**Regarding the design philosophy of `lint` vs `validate`:**

The distinction between local `lint` and server-side `validate` makes a lot of sense. The original intent of keeping `lint` as a pure local command is sound — it should be fast, offline, and deterministic.

The `adc lint` + `adc validate` + `adc sync` workflow you described is a cleaner separation of concerns:
- `lint`: local structural validation (Zod schemas)
- `validate`: server-side full validation (including plugin schemas + code-driven checks)
- `sync`: actual deployment

**Next steps:**

I'll defer to @nic-6443 on how to proceed. The options as I see them:
1. Close this PR and implement a proper `validate` subcommand instead
2. Refactor the plugin validation logic into a `validate` subcommand, keeping `lint` purely local
3. Continue discussion on the approach

cc @nic-6443 — please take a look at the reviewer's feedback and decide how you'd like to proceed.

**Comment by nic-6443** [comment]
> This does not solve 90% of the problems, because configuration errors outside of any plugins cannot be checked. 

@bzp2010 Why is that? Even though some plugins need to be validated through Lua code, most plugin configurations are validated via json schema. What exactly do you mean by the 90% of the problems you mentioned?

If you consider those plugins that need to be verified through Lua code, we can provide an API on the backend for the ADC to perform remote verification. Would this solve the problem 100%?

**Comment by bzp2010** [comment]
@nic-6443 

- [x] Resource configuration checks [the current `lint` tool does its best to do this, but no guarantee]
- [x] Plugin json schema checks [maybe]
- [ ] Plugin code checks
- [ ] Resource dependency checks, Etc.

Any issue will cause the synchronization process to be interrupted.

Local lint is useless against the latter two issues. Consequently, "false negatives" in the `lint` report will lead to "partial synchronization," which is effectively the same as missing synchronization at all. This results in unexpected behavior on the gateway (which I do not want to happen). No matter how many engineering resources we invest in the current approach, it is neither sustainable nor does it address the root cause of the problem.

This introduces unnecessary complexity.

> And without a once-and-for-all solution, I am unwilling to endorse the functionality claimed by the lint tool.

If the solution doesn’t fully resolve (validation passed = server unconditionally accepts the configuration during synchronization) the issue, I’d rather not do it. Engineering solutions can resolve the issue hacky.

**Comment by nic-6443** [comment]
@bzp2010 Is there any linter tool that claims to solve all production problems? All the user needs is a configuration checking capability, so why include all so-called synchronization failure issues in the current requirement?

**Comment by bzp2010** [comment]
If we don’t do it, the sync process might be interrupted by an error during synchronization.
If we do it, the sync process continues to be interrupted by errors.
So what’s the point of adding it?

**Comment by nic-6443** [comment]
Why is there a lint function now? Following this reasoning, wouldn't the lint function be completely worthless? But in fact, users are using it and hoping it will be more powerful.

**Comment by bzp2010** [comment]
You are right.

The fundamental issue isn’t whether an ADC can detect and flag errors locally, but rather that we can never fully align client-side and server-side checks.
To put it bluntly, this makes any attempt at local checking laughable and completely pointless. Lint merely performs the most basic checks using a set of predefined rules. Its implementation is very limited. It’s primarily used to catch basic issues like spelling mistakes.

UPDATE: The Zod schema does not allow any undefined fields to pass validation, so the spell check is entirely valid.

**Comment by bzp2010** [comment]
https://github.com/api7/adc/blob/b3ef55f27d8f3b872d7024cdb87beb59e9c889eb/apps/cli/src/command/lint.command.ts#L7-L11

Furthermore, if you read the code carefully, you’ll discover the original design intent behind the `lint` subcommand.

THAT IS `meets ADC requirements`.

It was never intended to ensure that the configuration passes server-side validation, but rather to verify that the input in the local configuration file meets ADC’s own formatting requirements—such as mandatory fields, correct data types, valid data ranges and spelling errors.

To elaborate, this is primarily to ensure that the subsequent ADC-to-Backend types conversion does not fail due to dirty data.

A predefined Zod schema handles this process quite well; while I cannot guarantee its absolute effectiveness, it is sufficient for most scenarios. Users can choose to opt out of this (by `--no-lint`), but I am not responsible for any program crashes that may result.

Yes, it's mainly to prevent program crashes.
